0023-executor-primitive.md

```diff
@@ -0,0 +1,188 @@
+noise learning, twirling, and expectation value calculations.
+
+This proposal will enable Qiskit users to perform large-scale, backend-optimized
+error mitigation experiments without incurring the high network cost of
+transmitting thousands or millions of circuit variations. By shifting variation
+generation to the backend through a portable DSL, researchers can run advanced
+techniques more efficiently.
+
+Today, Qiskit users can already run large-scale error mitigation experiments,
+but the process is **implicit**—it is difficult to explain and difficult to
+compose, and users have no direct control over how circuit variations are
+generated or how mitigation is applied. This proposal introduces the **samplex**
+DSL, a portable description language that lets users explicitly define the
+variation strategies applied by the backend.
+
+With this, researchers can tailor error mitigation to their needs. The approach
+preserves backend optimizations while giving users fine-grained
+control, reproducibility through deterministic seeds, and access to structured
+metadata for postprocessing.
+
+## User Benefit
+This proposal will primarily benefit researchers and practitioners working with
+noisy quantum hardware who require precise control over error mitigation
+strategies. By making the variation generation process explicit and
+user-definable, it empowers users to explore and tune novel error mitigation
+techniques.
+
+Backend and platform developers will also benefit, as the **samplex** DSL
+creates a standard interface for describing variation strategies that can be
+implemented consistently across vendors. This improves portability, reduces
+vendor lock-in, and enables backends to apply optimizations without sacrificing
+user intent.  
+
+Finally, the broader Qiskit community, including educators and tool developers,
+will gain a clearer and more composable model for error mitigation workflows,
+making it easier to experiment, reproduce results, and share techniques.
+
+## Design Proposal
+
+The intent of the following listing is to demonstrate how a user could leverage
+the new low-level interface to implement a client-side basic estimator.
+
+```python
+from qiskit_ibm_runtime import QiskitRuntimeService, Session, NoiseLearner
+from qiskit.transpiler import PassManager
+from qiskit.transpiler.passes import NoiseLearningLayering
+from qiskit.circuit import QuantumCircuit, Parameter, QuantumRegister, ClassicalRegister
+
+# This proposal assumes the semantics implemented in samplomatic will become
+# part of qiskit semantics at some moment.
+from qiskit import samplex
+from qiskit.samplex.annotations import InjectNoise, Twirl
+
+# The proposal wants to acknowledge the transtionary reality in which the
+# primitives continue being the entry point for operating the quantum computer
+# and this is a low level interface for advanced usage.
+from qiskit.primitives.implementation import QuantumProgram, Executor
```

Comment:
I expect that the `Executor` will need to come from `qiskit-ibm-runtime` instead: it needs to do things like POST to our service.

Instead, base class like BaseExecutor could live here.

I'm also wondering how this might all fit in with BackendV3, @ElePT , @jakelishman , @mtreinish : we might consider not having an `Executor` at all and reverting to something more like `backend.run()`. In this case we might need to have different semantics for sessions/batches than we have today.

Reply:
> I expect that the Executor will need to come from qiskit-ibm-runtime instead: it needs to do things like POST to our service.

They need to, same as sampler and estimator.

> I'm also wondering how this might all fit in with BackendV3, @ElePT , @jakelishman , @mtreinish : we might consider not having an Executor at all and reverting to something more like backend.run(). In this case we might need to have different semantics for sessions/batches than we have today.

I think this requires understanding how this proposal affects the "primitives model" of execution we currently have. Will this replace the primitives? complement the primitives? I don't know how we could reasonably justify reverting to `backend.run()` if the `sampler` and `estimator` primitives continue existing...

Reply:
Ok, I'm going to try to imitate the RFC for the [Estimator](https://github.com/Qiskit/RFCs/blob/master/0015-estimator-interface.md) in which the imports are not displayed, and elaborate on the concepts, then we can think of a strategy for adding this to Qiskit.